internal/btf/btf.go

```diff
@@ -78,21 +72,33 @@ func (h *btfHeader) stringStart() int64 {
 
 // LoadSpecFromReader reads from an ELF or a raw BTF blob.
 //
-// Returns ErrNotFound if reading from an ELF which contains no BTF.
-func LoadSpecFromReader(rd io.ReaderAt) (*Spec, error) {
+// Returns ErrNotFound if reading from an ELF which contains no BTF. ExtInfos
+// may be nil.
+func LoadSpecFromReader(rd io.ReaderAt) (*Spec, *ExtInfos, error) {
```

Comment:
I've seen you refer to eliminating `*ExtInfos` here in the future in your commit message, but can we get away with doing that sooner rather than later? Could we put `*ExtInfos` (back?) into the `btf.Spec` instead of returning it as a separate value here? IMO `LoadSpec*` should return a single value.

Reply:
I'd like to remove ExtInfos completely, but https://github.com/cilium/ebpf/pull/606#issuecomment-1104201245 is my best idea how to do it and I don't like it very much.

Embedding ExtInfos in Spec has it the wrong way round: when reading BTF we always have a Spec, but we don't always have ExtInfos. So modeling the relationship as ExtInfo{Spec} is more natural. For example, naked vmlinux BTF has no ext infos. Check the call sites of `loadRawSpec`, almost always we pass `nil, nil` and aren't interested in ExtInfos in the first place. With ExtInfos split out / removed a good amount of code can be refactored to be less deeply nested.

I was considering adding a separate constructor `LoadExtInfos(io.Reader)` and making `LoadSpec` skip ExtInfos. However I can't come up with a way to achieve that without (a) adding `*elf.File` to the `btf` API surface like `LoadExtInfos(*elf.File)` (and all the SafeELFFile baggage) or (b) parsing the ELF one more time. (We already parse the ELF header twice in the ELF loader due to this.)

Reply:
Well, check this out: https://github.com/cilium/ebpf/commit/640c4b0385e39775010880e92bd7711797e39e73 Kinda terrible, but we could use it to have `LoadExtInfosFromReader(io.ReaderAt)` and `LoadSpecFromReader(io.ReaderAt)`.

Reply:
Regarding ExtInfos: could we somehow break out that review and discussion into a separate PR? Taking `*elf.File` is not the way to go for the reasons you mentioned, but for me primarily due to #550.

If we're reworking the btf/extinfo reader API, we should definitely take the opportunity to address this current limitation and come to a solution that works for module BTF as well. Actually parsing (and 'linking') module BTF might be a bigger chunk of work, but at least the API should allow for it to be added later if needed.